### PR TITLE
config: add no delay option for tcp connection

### PR DESCRIPTION
**Pull request by jianzhiyao**
### What problem does this PR solve?

Add option to reduce tcp packets number.

### What is changed and how it works?

What's Changed:

Add ``TCPNoDelay`` option into struct of Performance. And call ``net.TCPConn.SetNoDelay(false)`` when it is ``false``

How it Works:

``net.TCPConn.SetNoDelay(false)`` can make tcp packet size as big as possible and make tcp packets number as less as possible, it will be useful in a network with bad situation

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below) 

#### sysbench: TCPNoDelay == true
```
> docker run -t --rm dotnetdr/sysbench:0.5 sysbench --num-threads=12 --max-time=60 --max-requests=999999999 --test=sysbench/tests/db/oltp.lua --oltp-table-size=100000 --oltp-tables-count=3 --db-driver=mysql --mysql-table-engine=innodb --mysql-host=host.docker.internal --mysql-port=4000 --mysql-user=root  --mysql-db=aaa prepare && \ 
> docker run -t dotnetdr/sysbench:0.5 sysbench --num-threads=1 --max-time=60 --max-requests=999999999 --test=sysbench/tests/db/oltp.lua --oltp-table-size=100000 --oltp-tables-count=3 --db-driver=mysql --mysql-table-engine=innodb --mysql-host=host.docker.internal --mysql-port=4000 --mysql-user=root  --mysql-db=aaa --report-interval=60 run
[  60s] threads: 1, tps: 32.58, reads: 456.23, writes: 130.34, response time: 38.46ms (95%), errors: 0.00, reconnects:  0.00
OLTP test statistics:
    queries performed:
        read:                            27384
        write:                           7824
        other:                           3912
        total:                           39120
    transactions:                        1956   (32.59 per sec.)
    read/write requests:                 35208  (586.55 per sec.)
    other operations:                    3912   (65.17 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          60.0252s
    total number of events:              1956
    total time taken by event execution: 60.0067s
    response time:
         min:                                 23.28ms
         avg:                                 30.68ms
         max:                                 75.12ms
         approx.  95 percentile:              38.46ms

Threads fairness:
    events (avg/stddev):           1956.0000/0.00
    execution time (avg/stddev):   60.0067/0.00
```

#### sysbench: TCPNoDelay == false
```
> docker run -t --rm dotnetdr/sysbench:0.5 sysbench --num-threads=12 --max-time=60 --max-requests=999999999 --test=sysbench/tests/db/oltp.lua --oltp-table-size=100000 --oltp-tables-count=3 --db-driver=mysql --mysql-table-engine=innodb --mysql-host=host.docker.internal --mysql-port=4000 --mysql-user=root  --mysql-db=bbb prepare && \ 
> docker run -t dotnetdr/sysbench:0.5 sysbench --num-threads=1 --max-time=60 --max-requests=999999999 --test=sysbench/tests/db/oltp.lua --oltp-table-size=100000 --oltp-tables-count=3 --db-driver=mysql --mysql-table-engine=innodb --mysql-host=host.docker.internal --mysql-port=4000 --mysql-user=root  --mysql-db=bbb --report-interval=60 run
[  60s] threads: 1, tps: 33.88, reads: 474.55, writes: 135.53, response time: 34.86ms (95%), errors: 0.00, reconnects:  0.00
OLTP test statistics:
    queries performed:
        read:                            28476
        write:                           8136
        other:                           4068
        total:                           40680
    transactions:                        2034   (33.89 per sec.)
    read/write requests:                 36612  (610.02 per sec.)
    other operations:                    4068   (67.78 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          60.0174s
    total number of events:              2034
    total time taken by event execution: 60.0025s
    response time:
         min:                                 22.82ms
         avg:                                 29.50ms
         max:                                123.35ms
         approx.  95 percentile:              34.88ms

Threads fairness:
    events (avg/stddev):           2034.0000/0.00
    execution time (avg/stddev):   60.0025/0.00
```

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note

- No release note

### CODE REVIEW

**Comment by zz-jason** [comment]
Is it possible to get a performance after setting `tcp-no-delay` to `false` in a certain scenario?

**Comment by jianzhiyao** [comment]
According to [mysql doc](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-connp-props-networking.html) and [Best Practices for TCP Optimization in 2019](https://www.extrahop.com/company/blog/2016/tcp-nodelay-nagle-quickack-best-practices/), i think using ``Nagle's algorithm`` is useful for user in some specific situation.
So,I think tidb should allow user have a choice for this option like mysql.

File: config/config.go [hunk 1]
```diff
@@ -593,6 +594,7 @@ var defaultConf = Config{
 		ServerMemoryQuota:     0,
 		MemoryUsageAlarmRatio: 0.8,
 		TCPKeepAlive:          true,
+		TCPNoDelay:            true,
```
Comment (xhebox) [inline config/config.go:597]: It needs more internal decisions on the default state of this switch. But if it defaults to false, I think we could merge this PR and you don't need to set it to false in the test.

**Review by xhebox — COMMENTED** [review]
/cc @djshow832 @bb7133 a small feature, i guess we need an issue to track the problem and the docs

File: config/config.go (hunk 1 above)
Comment (jianzhiyao) [reply config/config.go:597]: According to https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-connp-props-networking.html, TCPNoDelay is positive as default.

Comment (xhebox) [reply config/config.go:597]: I saw that in your comment. I agreed that it is effective. But the decision may not be determined in a short time... I would suggest you merge the PR with an default-off-switch first: at least you will have an option that can be enabled.

Join the slack channel #sig-sql-infra for discussion with members.

**Review by djshow832 — COMMENTED** [review]
Can you add a simple test case for this config?

**Comment by jianzhiyao** [comment]
PTAL @djshow832

**Comment by jianzhiyao** [comment]
PTAL @xhebox

**Comment by xhebox** [comment]
/lgtm

@jiangyuzhao You may want to update pingcap/docs and pingcap/docs-cn, the tidb-configuration-file.md file.

**Pull request merged**